File: swift-bridge/Sources/SpriteKitBridge/PhysicsJoint.swift
import SpriteKit

@_cdecl("sk_physics_joint_get_reaction_force_dx")
public func sk_physics_joint_get_reaction_force_dx(_ jointHandle: UnsafeMutableRawPointer?) -> Double {
    guard let joint: SKPhysicsJoint = skBorrow(jointHandle) else { return 0 }
    return Double(joint.reactionForce.dx)
}

@_cdecl("sk_physics_joint_get_reaction_force_dy")
public func sk_physics_joint_get_reaction_force_dy(_ jointHandle: UnsafeMutableRawPointer?) -> Double {
    guard let joint: SKPhysicsJoint = skBorrow(jointHandle) else { return 0 }
    return Double(joint.reactionForce.dy)
}

@_cdecl("sk_physics_joint_get_reaction_torque")
public func sk_physics_joint_get_reaction_torque(_ jointHandle: UnsafeMutableRawPointer?) -> Double {
    guard let joint: SKPhysicsJoint = skBorrow(jointHandle) else { return 0 }
    return Double(joint.reactionTorque)
}

@_cdecl("sk_physics_joint_pin_new")
public func sk_physics_joint_pin_new(_ bodyAHandle: UnsafeMutableRawPointer?, _ bodyBHandle: UnsafeMutableRawPointer?, _ x: Double, _ y: Double) -> UnsafeMutableRawPointer? {
    guard let bodyA: SKPhysicsBody = skBorrow(bodyAHandle),
          let bodyB: SKPhysicsBody = skBorrow(bodyBHandle)
    else { return nil }
    return skRetain(SKPhysicsJointPin.joint(withBodyA: bodyA, bodyB: bodyB, anchor: CGPoint(x: x, y: y)))
}

@_cdecl("sk_physics_joint_pin_get_should_enable_limits")
public func sk_physics_joint_pin_get_should_enable_limits(_ jointHandle: UnsafeMutableRawPointer?) -> Bool {
    guard let joint: SKPhysicsJointPin = skBorrow(jointHandle) else { return false }
    return joint.shouldEnableLimits
}

@_cdecl("sk_physics_joint_pin_set_should_enable_limits")
public func sk_physics_joint_pin_set_should_enable_limits(_ jointHandle: UnsafeMutableRawPointer?, _ enabled: Bool) {
    guard let joint: SKPhysicsJointPin = skBorrow(jointHandle) else { return }
    joint.shouldEnableLimits = enabled
}

@_cdecl("sk_physics_joint_pin_get_lower_angle_limit")
public func sk_physics_joint_pin_get_lower_angle_limit(_ jointHandle: UnsafeMutableRawPointer?) -> Double {
    guard let joint: SKPhysicsJointPin = skBorrow(jointHandle) else { return 0 }
    return Double(joint.lowerAngleLimit)
}

@_cdecl("sk_physics_joint_pin_set_lower_angle_limit")
public func sk_physics_joint_pin_set_lower_angle_limit(_ jointHandle: UnsafeMutableRawPointer?, _ value: Double) {
    guard let joint: SKPhysicsJointPin = skBorrow(jointHandle) else { return }
    joint.lowerAngleLimit = CGFloat(value)
}

@_cdecl("sk_physics_joint_pin_get_upper_angle_limit")
public func sk_physics_joint_pin_get_upper_angle_limit(_ jointHandle: UnsafeMutableRawPointer?) -> Double {
    guard let joint: SKPhysicsJointPin = skBorrow(jointHandle) else { return 0 }
    return Double(joint.upperAngleLimit)
}

@_cdecl("sk_physics_joint_pin_set_upper_angle_limit")
public func sk_physics_joint_pin_set_upper_angle_limit(_ jointHandle: UnsafeMutableRawPointer?, _ value: Double) {
    guard let joint: SKPhysicsJointPin = skBorrow(jointHandle) else { return }
    joint.upperAngleLimit = CGFloat(value)
}

@_cdecl("sk_physics_joint_pin_get_friction_torque")
public func sk_physics_joint_pin_get_friction_torque(_ jointHandle: UnsafeMutableRawPointer?) -> Double {
    guard let joint: SKPhysicsJointPin = skBorrow(jointHandle) else { return 0 }
    return Double(joint.frictionTorque)
}

@_cdecl("sk_physics_joint_pin_set_friction_torque")
public func sk_physics_joint_pin_set_friction_torque(_ jointHandle: UnsafeMutableRawPointer?, _ value: Double) {
    guard let joint: SKPhysicsJointPin = skBorrow(jointHandle) else { return }
    joint.frictionTorque = CGFloat(value)
}

@_cdecl("sk_physics_joint_pin_get_rotation_speed")
public func sk_physics_joint_pin_get_rotation_speed(_ jointHandle: UnsafeMutableRawPointer?) -> Double {
    guard let joint: SKPhysicsJointPin = skBorrow(jointHandle) else { return 0 }
    return Double(joint.rotationSpeed)
}

@_cdecl("sk_physics_joint_pin_set_rotation_speed")
public func sk_physics_joint_pin_set_rotation_speed(_ jointHandle: UnsafeMutableRawPointer?, _ value: Double) {
    guard let joint: SKPhysicsJointPin = skBorrow(jointHandle) else { return }
    joint.rotationSpeed = CGFloat(value)
}

@_cdecl("sk_physics_joint_spring_new")
public func sk_physics_joint_spring_new(_ bodyAHandle: UnsafeMutableRawPointer?, _ bodyBHandle: UnsafeMutableRawPointer?, _ anchorAX: Double, _ anchorAY: Double, _ anchorBX: Double, _ anchorBY: Double) -> UnsafeMutableRawPointer? {
    guard let bodyA: SKPhysicsBody = skBorrow(bodyAHandle),
          let bodyB: SKPhysicsBody = skBorrow(bodyBHandle)
    else { return nil }
    return skRetain(SKPhysicsJointSpring.joint(withBodyA: bodyA, bodyB: bodyB, anchorA: CGPoint(x: anchorAX, y: anchorAY), anchorB: CGPoint(x: anchorBX, y: anchorBY)))
}

@_cdecl("sk_physics_joint_spring_get_damping")
public func sk_physics_joint_spring_get_damping(_ jointHandle: UnsafeMutableRawPointer?) -> Double {
    guard let joint: SKPhysicsJointSpring = skBorrow(jointHandle) else { return 0 }
    return Double(joint.damping)
}

@_cdecl("sk_physics_joint_spring_set_damping")
public func sk_physics_joint_spring_set_damping(_ jointHandle: UnsafeMutableRawPointer?, _ value: Double) {
    guard let joint: SKPhysicsJointSpring = skBorrow(jointHandle) else { return }
    joint.damping = CGFloat(value)
}

@_cdecl("sk_physics_joint_spring_get_frequency")
public func sk_physics_joint_spring_get_frequency(_ jointHandle: UnsafeMutableRawPointer?) -> Double {
    guard let joint: SKPhysicsJointSpring = skBorrow(jointHandle) else { return 0 }
    return Double(joint.frequency)
}

@_cdecl("sk_physics_joint_spring_set_frequency")
public func sk_physics_joint_spring_set_frequency(_ jointHandle: UnsafeMutableRawPointer?, _ value: Double) {
    guard let joint: SKPhysicsJointSpring = skBorrow(jointHandle) else { return }
    joint.frequency = CGFloat(value)
}

@_cdecl("sk_physics_joint_fixed_new")
public func sk_physics_joint_fixed_new(_ bodyAHandle: UnsafeMutableRawPointer?, _ bodyBHandle: UnsafeMutableRawPointer?, _ x: Double, _ y: Double) -> UnsafeMutableRawPointer? {
    guard let bodyA: SKPhysicsBody = skBorrow(bodyAHandle),
          let bodyB: SKPhysicsBody = skBorrow(bodyBHandle)
    else { return nil }
    return skRetain(SKPhysicsJointFixed.joint(withBodyA: bodyA, bodyB: bodyB, anchor: CGPoint(x: x, y: y)))
}

@_cdecl("sk_physics_joint_sliding_new")
public func sk_physics_joint_sliding_new(_ bodyAHandle: UnsafeMutableRawPointer?, _ bodyBHandle: UnsafeMutableRawPointer?, _ anchorX: Double, _ anchorY: Double, _ axisDX: Double, _ axisDY: Double) -> UnsafeMutableRawPointer? {
    guard let bodyA: SKPhysicsBody = skBorrow(bodyAHandle),
          let bodyB: SKPhysicsBody = skBorrow(bodyBHandle)
    else { return nil }
    return skRetain(SKPhysicsJointSliding.joint(withBodyA: bodyA, bodyB: bodyB, anchor: CGPoint(x: anchorX, y: anchorY), axis: CGVector(dx: axisDX, dy: axisDY)))
}

@_cdecl("sk_physics_joint_sliding_get_should_enable_limits")
public func sk_physics_joint_sliding_get_should_enable_limits(_ jointHandle: UnsafeMutableRawPointer?) -> Bool {
    guard let joint: SKPhysicsJointSliding = skBorrow(jointHandle) else { return false }
    return joint.shouldEnableLimits
}

@_cdecl("sk_physics_joint_sliding_set_should_enable_limits")
public func sk_physics_joint_sliding_set_should_enable_limits(_ jointHandle: UnsafeMutableRawPointer?, _ enabled: Bool) {
    guard let joint: SKPhysicsJointSliding = skBorrow(jointHandle) else { return }
    joint.shouldEnableLimits = enabled
}

@_cdecl("sk_physics_joint_sliding_get_lower_distance_limit")
public func sk_physics_joint_sliding_get_lower_distance_limit(_ jointHandle: UnsafeMutableRawPointer?) -> Double {
    guard let joint: SKPhysicsJointSliding = skBorrow(jointHandle) else { return 0 }
    return Double(joint.lowerDistanceLimit)
}

@_cdecl("sk_physics_joint_sliding_set_lower_distance_limit")
public func sk_physics_joint_sliding_set_lower_distance_limit(_ jointHandle: UnsafeMutableRawPointer?, _ value: Double) {
    guard let joint: SKPhysicsJointSliding = skBorrow(jointHandle) else { return }
    joint.lowerDistanceLimit = CGFloat(value)
}

@_cdecl("sk_physics_joint_sliding_get_upper_distance_limit")
public func sk_physics_joint_sliding_get_upper_distance_limit(_ jointHandle: UnsafeMutableRawPointer?) -> Double {
    guard let joint: SKPhysicsJointSliding = skBorrow(jointHandle) else { return 0 }
    return Double(joint.upperDistanceLimit)
}

@_cdecl("sk_physics_joint_sliding_set_upper_distance_limit")
public func sk_physics_joint_sliding_set_upper_distance_limit(_ jointHandle: UnsafeMutableRawPointer?, _ value: Double) {
    guard let joint: SKPhysicsJointSliding = skBorrow(jointHandle) else { return }
    joint.upperDistanceLimit = CGFloat(value)
}

@_cdecl("sk_physics_joint_limit_new")
public func sk_physics_joint_limit_new(_ bodyAHandle: UnsafeMutableRawPointer?, _ bodyBHandle: UnsafeMutableRawPointer?, _ anchorAX: Double, _ anchorAY: Double, _ anchorBX: Double, _ anchorBY: Double) -> UnsafeMutableRawPointer? {
    guard let bodyA: SKPhysicsBody = skBorrow(bodyAHandle),
          let bodyB: SKPhysicsBody = skBorrow(bodyBHandle)
    else { return nil }
    return skRetain(SKPhysicsJointLimit.joint(withBodyA: bodyA, bodyB: bodyB, anchorA: CGPoint(x: anchorAX, y: anchorAY), anchorB: CGPoint(x: anchorBX, y: anchorBY)))
}

@_cdecl("sk_physics_joint_limit_get_max_length")
public func sk_physics_joint_limit_get_max_length(_ jointHandle: UnsafeMutableRawPointer?) -> Double {
    guard let joint: SKPhysicsJointLimit = skBorrow(jointHandle) else { return 0 }
    return Double(joint.maxLength)
}

@_cdecl("sk_physics_joint_limit_set_max_length")
public func sk_physics_joint_limit_set_max_length(_ jointHandle: UnsafeMutableRawPointer?, _ value: Double) {
    guard let joint: SKPhysicsJointLimit = skBorrow(jointHandle) else { return }
    joint.maxLength = CGFloat(value)
}

File: src/physics_joint.rs
use apple_cf::cg::{CGPoint, CGVector};

use crate::ffi;
use crate::physics_body::PhysicsBody;
use crate::private::handle_type;

/// Trait for `SKPhysicsJoint`.
pub trait AsPhysicsJoint {
    #[doc(hidden)]
    fn as_joint_ptr(&self) -> *mut core::ffi::c_void;
}

/// Trait for `SKPhysicsJoint`.
pub trait PhysicsJointExt: AsPhysicsJoint {
    #[must_use]
    fn reaction_force(&self) -> CGVector {
        CGVector::new(
            unsafe { ffi::sk_physics_joint_get_reaction_force_dx(self.as_joint_ptr()) },
            unsafe { ffi::sk_physics_joint_get_reaction_force_dy(self.as_joint_ptr()) },
        )
    }

    #[must_use]
    fn reaction_torque(&self) -> f64 {
        unsafe { ffi::sk_physics_joint_get_reaction_torque(self.as_joint_ptr()) }
    }
}

impl<T: AsPhysicsJoint> PhysicsJointExt for T {}

handle_type!(PhysicsJointPin);
handle_type!(PhysicsJointSpring);
handle_type!(PhysicsJointFixed);
handle_type!(PhysicsJointSliding);
handle_type!(PhysicsJointLimit);

impl AsPhysicsJoint for PhysicsJointPin {
    fn as_joint_ptr(&self) -> *mut core::ffi::c_void {
        self.ptr
    }
}

impl AsPhysicsJoint for PhysicsJointSpring {
    fn as_joint_ptr(&self) -> *mut core::ffi::c_void {
        self.ptr
    }
}

impl AsPhysicsJoint for PhysicsJointFixed {
    fn as_joint_ptr(&self) -> *mut core::ffi::c_void {
        self.ptr
    }
}

impl AsPhysicsJoint for PhysicsJointSliding {
    fn as_joint_ptr(&self) -> *mut core::ffi::c_void {
        self.ptr
    }
}

impl AsPhysicsJoint for PhysicsJointLimit {
    fn as_joint_ptr(&self) -> *mut core::ffi::c_void {
        self.ptr
    }
}

impl PhysicsJointPin {
    /// Wraps `SKPhysicsJointPin`.
    #[must_use]
    pub fn new(body_a: &PhysicsBody, body_b: &PhysicsBody, anchor: CGPoint) -> Option<Self> {
        unsafe {
            Self::from_raw(ffi::sk_physics_joint_pin_new(
                body_a.as_ptr(),
                body_b.as_ptr(),
                anchor.x,
                anchor.y,
            ))
        }
    }

    /// Returns a property exposed by `SKPhysicsJointPin`.
    #[must_use]
    pub fn should_enable_limits(&self) -> bool {
        unsafe { ffi::sk_physics_joint_pin_get_should_enable_limits(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsJointPin`.
    pub fn set_should_enable_limits(&self, enabled: bool) {
        unsafe { ffi::sk_physics_joint_pin_set_should_enable_limits(self.ptr, enabled) };
    }

    /// Returns a property exposed by `SKPhysicsJointPin`.
    #[must_use]
    pub fn lower_angle_limit(&self) -> f64 {
        unsafe { ffi::sk_physics_joint_pin_get_lower_angle_limit(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsJointPin`.
    pub fn set_lower_angle_limit(&self, value: f64) {
        unsafe { ffi::sk_physics_joint_pin_set_lower_angle_limit(self.ptr, value) };
    }

    /// Returns a property exposed by `SKPhysicsJointPin`.
    #[must_use]
    pub fn upper_angle_limit(&self) -> f64 {
        unsafe { ffi::sk_physics_joint_pin_get_upper_angle_limit(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsJointPin`.
    pub fn set_upper_angle_limit(&self, value: f64) {
        unsafe { ffi::sk_physics_joint_pin_set_upper_angle_limit(self.ptr, value) };
    }

    /// Returns a property exposed by `SKPhysicsJointPin`.
    #[must_use]
    pub fn friction_torque(&self) -> f64 {
        unsafe { ffi::sk_physics_joint_pin_get_friction_torque(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsJointPin`.
    pub fn set_friction_torque(&self, value: f64) {
        unsafe { ffi::sk_physics_joint_pin_set_friction_torque(self.ptr, value) };
    }

    /// Returns a property exposed by `SKPhysicsJointPin`.
    #[must_use]
    pub fn rotation_speed(&self) -> f64 {
        unsafe { ffi::sk_physics_joint_pin_get_rotation_speed(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsJointPin`.
    pub fn set_rotation_speed(&self, value: f64) {
        unsafe { ffi::sk_physics_joint_pin_set_rotation_speed(self.ptr, value) };
    }
}

impl PhysicsJointSpring {
    /// Wraps `SKPhysicsJointSpring`.
    #[must_use]
    pub fn new(
        body_a: &PhysicsBody,
        body_b: &PhysicsBody,
        anchor_a: CGPoint,
        anchor_b: CGPoint,
    ) -> Option<Self> {
        unsafe {
            Self::from_raw(ffi::sk_physics_joint_spring_new(
                body_a.as_ptr(),
                body_b.as_ptr(),
                anchor_a.x,
                anchor_a.y,
                anchor_b.x,
                anchor_b.y,
            ))
        }
    }

    /// Returns a property exposed by `SKPhysicsJointSpring`.
    #[must_use]
    pub fn damping(&self) -> f64 {
        unsafe { ffi::sk_physics_joint_spring_get_damping(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsJointSpring`.
    pub fn set_damping(&self, value: f64) {
        unsafe { ffi::sk_physics_joint_spring_set_damping(self.ptr, value) };
    }

    /// Returns a property exposed by `SKPhysicsJointSpring`.
    #[must_use]
    pub fn frequency(&self) -> f64 {
        unsafe { ffi::sk_physics_joint_spring_get_frequency(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsJointSpring`.
    pub fn set_frequency(&self, value: f64) {
        unsafe { ffi::sk_physics_joint_spring_set_frequency(self.ptr, value) };
    }
}

impl PhysicsJointFixed {
    /// Wraps `SKPhysicsJointFixed`.
    #[must_use]
    pub fn new(body_a: &PhysicsBody, body_b: &PhysicsBody, anchor: CGPoint) -> Option<Self> {
        unsafe {
            Self::from_raw(ffi::sk_physics_joint_fixed_new(
                body_a.as_ptr(),
                body_b.as_ptr(),
                anchor.x,
                anchor.y,
            ))
        }
    }
}

impl PhysicsJointSliding {
    /// Wraps `SKPhysicsJointSliding`.
    #[must_use]
    pub fn new(
        body_a: &PhysicsBody,
        body_b: &PhysicsBody,
        anchor: CGPoint,
        axis: CGVector,
    ) -> Option<Self> {
        unsafe {
            Self::from_raw(ffi::sk_physics_joint_sliding_new(
                body_a.as_ptr(),
                body_b.as_ptr(),
                anchor.x,
                anchor.y,
                axis.dx,
                axis.dy,
            ))
        }
    }

    /// Returns a property exposed by `SKPhysicsJointSliding`.
    #[must_use]
    pub fn should_enable_limits(&self) -> bool {
        unsafe { ffi::sk_physics_joint_sliding_get_should_enable_limits(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsJointSliding`.
    pub fn set_should_enable_limits(&self, enabled: bool) {
        unsafe { ffi::sk_physics_joint_sliding_set_should_enable_limits(self.ptr, enabled) };
    }

    /// Returns a property exposed by `SKPhysicsJointSliding`.
    #[must_use]
    pub fn lower_distance_limit(&self) -> f64 {
        unsafe { ffi::sk_physics_joint_sliding_get_lower_distance_limit(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsJointSliding`.
    pub fn set_lower_distance_limit(&self, value: f64) {
        unsafe { ffi::sk_physics_joint_sliding_set_lower_distance_limit(self.ptr, value) };
    }

    /// Returns a property exposed by `SKPhysicsJointSliding`.
    #[must_use]
    pub fn upper_distance_limit(&self) -> f64 {
        unsafe { ffi::sk_physics_joint_sliding_get_upper_distance_limit(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsJointSliding`.
    pub fn set_upper_distance_limit(&self, value: f64) {
        unsafe { ffi::sk_physics_joint_sliding_set_upper_distance_limit(self.ptr, value) };
    }
}

impl PhysicsJointLimit {
    /// Wraps `SKPhysicsJointLimit`.
    #[must_use]
    pub fn new(
        body_a: &PhysicsBody,
        body_b: &PhysicsBody,
        anchor_a: CGPoint,
        anchor_b: CGPoint,
    ) -> Option<Self> {
        unsafe {
            Self::from_raw(ffi::sk_physics_joint_limit_new(
                body_a.as_ptr(),
                body_b.as_ptr(),
                anchor_a.x,
                anchor_a.y,
                anchor_b.x,
                anchor_b.y,
            ))
        }
    }

    /// Returns a property exposed by `SKPhysicsJointLimit`.
    #[must_use]
    pub fn max_length(&self) -> f64 {
        unsafe { ffi::sk_physics_joint_limit_get_max_length(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsJointLimit`.
    pub fn set_max_length(&self, value: f64) {
        unsafe { ffi::sk_physics_joint_limit_set_max_length(self.ptr, value) };
    }
}
